File: web/src/api/survey.ts
import { api } from './client'

export interface SurveyStatus {
  request_id: string
  token_status: string
  expired: boolean
  opened_at: string | null
  verified_at: string | null
  submitted_at: string | null
  hospital_name: string | null
  doctor_name: string | null
  patient_name_masked: string | null
  timepoint_label: string | null
}

export interface SurveyVerifyRequest {
  method_code: 'birth_ymd' | 'phone_last4'
  value: string
}

export interface SurveyVerifyResponse {
  verified: boolean
  survey_token: string
  expires_at: string
}

export interface SurveyQuestion {
  step: number
  category: string
  instrument_code: string
  question_code: string
  title: string
  ui_type: string
  options: { value: number | string; label: string }[] | null
  min_val: number | null
  max_val: number | null
}

export interface SurveyQuestionsResponse {
  patient_name: string | null
  doctor_name: string | null
  timepoint_label: string | null
  total_questions: number
  questions: SurveyQuestion[]
}

export interface SurveySubmitResponse {
  is_completed: boolean
  sync_status: string
  submitted_at: string
}

export const surveyService = {
  /** Get survey status (no auth required) */
  getStatus: (tokenUuid: string) =>
    api.get<SurveyStatus>(`/survey/${tokenUuid}/status`),

  /** Verify patient identity */
  verify: (tokenUuid: string, data: SurveyVerifyRequest) =>
    api.post<SurveyVerifyResponse>(`/survey/${tokenUuid}/verify`, data),

  /** Fetch survey questions (requires survey token) */
  getQuestions: (tokenUuid: string, surveyToken: string) =>
    api.get<SurveyQuestionsResponse>(`/survey/${tokenUuid}/questions`, surveyToken),

  /** Save draft answer */
  saveDraft: (tokenUuid: string, data: { question_id: string; answer_value: unknown }, surveyToken: string) =>
    api.patch<void>(`/survey/${tokenUuid}/save`, data, surveyToken),

  /** Submit all answers */
  submit: (tokenUuid: string, answers: Record<string, unknown>, surveyToken: string) =>
    api.post<SurveySubmitResponse>(`/survey/${tokenUuid}/submit`, { answers }, surveyToken),
}

File: server/app/models/survey.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import Field

from app.models.common import APIModel


class SurveyStatusResponse(APIModel):
    request_id: UUID
    token_status: str
    expired: bool
    opened_at: datetime | None = None
    verified_at: datetime | None = None
    submitted_at: datetime | None = None
    hospital_name: str | None = None
    doctor_name: str | None = None
    patient_name_masked: str | None = None
    timepoint_label: str | None = None


class SurveyVerifyRequest(APIModel):
    method_code: str = Field(pattern="^(birth_ymd|phone_last4)$")
    value: str = Field(min_length=4, max_length=20)


class SurveyVerifyResponse(APIModel):
    verified: bool
    survey_token: str
    expires_at: datetime


class SurveyQuestionOption(APIModel):
    value: int | str
    label: str


class SurveyQuestion(APIModel):
    step: int
    category: str
    instrument_code: str
    question_code: str
    title: str
    ui_type: str
    options: list[SurveyQuestionOption] | None = None
    min_val: int | None = None
    max_val: int | None = None


class SurveyQuestionsResponse(APIModel):
    patient_name: str | None = None
    doctor_name: str | None = None
    timepoint_label: str | None = None
    total_questions: int
    questions: list[SurveyQuestion]


class SurveySaveRequest(APIModel):
    question_id: str
    answer_value: int | str | bool | float | dict | list | None


class SurveySubmitRequest(APIModel):
    answers: dict[str, int | str | bool | float | dict | list | None]


class SurveySubmitResponse(APIModel):
    is_completed: bool
    sync_status: str = "SUCCESS"
    submitted_at: datetime
